src/main/ruby/core/posix.rb

```diff
@@ -42,6 +42,14 @@ def self.to_nfi_type(type)
     end
   end
 
+  def self.attach_function_lazily(native_name, argument_types, return_type,
+                                  on: Truffle::POSIX, as: native_name,
+                                  library: LIBC, blocking: false)
```

Comment:
The signature could be much simpler here, just `*attach_args`.

Reply:
I disagree, that would force reader to investigate the body of the method to see the `attach_function` signature to get the actual arguments. It also breaks IntelliJ method arguments hints.

Reply:
But it does duplicate code, makes it harder to change the method below, has a cost on performance, etc.
TBH I don't care about IntelliJ support of arguments for this, the examples below are anyway more useful than what an IDE could provide.

Reply:
The one true advantage is that it checks the arguments at definition time and that's valuable.

Reply:
I don't think method signature can be counted as code duplication. Ruby has no tools to abstract it and reuse, `*attach_args` is omission not abstraction.

Yes that's true the editor needs to change 2 signatures not just one, however source code is read more often then edited, so the focus should be on the reader in this case. 

I don't think the performance impact would be even measurable in this case. We should rather focus on identifying real bottlenecks then doing this type of micro-optimizations.

You may not find it useful, but I am not the only one using idea so the following is useful
<img width="615" alt="screen shot 2017-11-14 at 14 42 14" src="https://user-images.githubusercontent.com/204460/32782956-6a2200da-c94a-11e7-9543-64b229d37cba.png">
not just to me but to others as well.

Yeah that's another advantage of keeping the whole signature along with my original argument that a reader should not be forced to read the body of the method to see how it's suppose to be called.

I remain convinced that the whole signature is better in this case.

Reply:
> I don't think the performance impact would be even measurable in this case. We should rather focus on identifying real bottlenecks then doing this type of micro-optimizations.

I am not so sure, Hash handling during startup can take a significant time, it would be good to measure if it makes a difference when loading the entire `posix.rb` file.

I'm fine to keep the signature (nicer to catch errors early), if the overhead is not too big.

Reply:
I've never seen IntelliJ parameter hints like that. IntelliJ thinks even basic things like `alias_method` are undefined for me so I'm not sure if it's broken somehow.

Reply:
I've configured my TruffleRuby repository as a SDK ruby in idea, and at the same time used it as an SDK for the TruffleRuby project. Then it works quite good (at lest for the methods defined in Ruby, `alias_method` is probably not one of them though).

Reply:
I like the IDE support. But if we just make `attach_function` lazy by default, I don't think we need the delegated call at all and this issue goes away.

Reply:
To clarify, I'm OK with duplicating arguments here.
I'll evaluate if it has an impact on startup time, but that can be done separately from this PR.